rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2226 @@
+    where
+        Src: TransmuteInto<Dst, Neglect>,
+        Neglect: SafeTransmuteOptions
+    {
+        src.transmute_into()
+    }
+
+    /// Reinterprets the bits of a value of one type as another type, potentially unsafely.
+    #[inline(always)]
+    pub /*const*/ unsafe fn unsafe_transmute<Src, Dst, Neglect>(src: Src) -> Dst
+    where
+        Src: TransmuteInto<Dst, Neglect>,
+        Neglect: TransmuteOptions
+    {
+        unsafe { src.unsafe_transmute_into() }
+    }
+
+    /// `Self: TransmuteInto<Dst, Neglect`, if the compiler accepts the stability,
+    /// safety, and soundness of transmuting `Self` into `Dst`, notwithstanding
+    /// a given set of static checks to `Neglect`.
+    pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
+    where
+        Neglect: TransmuteOptions,
+    {
+        /// Reinterpret the bits of a value of one type as another type, safely.
+        fn transmute_into(self) -> Dst
+        where
+            Self: Sized,
+            Dst: Sized,
+            Neglect: SafeTransmuteOptions;
+
+        /// Reinterpret the bits of a value of one type as another type, potentially unsafely.
+        ///
+        /// The onus is on you to ensure that calling this method is safe.
+        unsafe fn unsafe_transmute_into(self) -> Dst
+        where
+            Self: Sized,
+            Dst: Sized,
+            Neglect: TransmuteOptions;
+    }
+
+    unsafe impl<Src, Dst, Neglect> TransmuteInto<Dst, Neglect> for Src
+    where
+        Src: ?Sized,
+        Dst: ?Sized + TransmuteFrom<Src, Neglect>,
+        Neglect: TransmuteOptions,
+    {
+        #[inline(always)]
+        fn transmute_into(self) -> Dst
+        where
+            Self: Sized,
+            Dst: Sized,
+            Neglect: SafeTransmuteOptions,
+        {
+            Dst::transmute_from(self)
+        }
+
+        #[inline(always)]
+        unsafe fn unsafe_transmute_into(self) -> Dst
+        where
+            Self: Sized,
+            Dst: Sized,
+            Neglect: TransmuteOptions,
+        {
+            unsafe { Dst::unsafe_transmute_from(self) }
+        }
+    }
+
+    /// `Self: TransmuteInto<Src, Neglect`, if the compiler accepts the stability,
+    /// safety, and soundness of transmuting `Src` into `Self`, notwithstanding
+    /// a given set of static checks to `Neglect`.
+    /* #[lang = "transmute_from"] */
+    pub unsafe trait TransmuteFrom<Src: ?Sized, Neglect = ()>
+    where
+        Neglect: TransmuteOptions,
+    {
+        /// Reinterpret the bits of a value of one type as another type, safely.
+        #[inline(always)]
+        fn transmute_from(src: Src) -> Self
+        where
+            Src: Sized,
+            Self: Sized,
+            Neglect: SafeTransmuteOptions,
+        {
+            use core::{mem, ptr};
+            unsafe {
+                let dst = ptr::read(&src as *const Src as *const Self);
+                mem::forget(src);
+                dst
+            }
```

Comment:
This violates Stacked Borrows when `Src` is a type that asserts exclusivity (_e.g._, `Src = Box<T>, Dst = &'any T`).

```suggestion
            unsafe {
                let src_no_drop = ManuallyDrop::new(src);
                let dst = ptr::read(&*src_no_drop as *const Src as *const Self);
                dst
            }
```
(or `ManuallyDrop::take`, or `transmute_copy`, but these one do not generalize to the `read_unaligned` cases below)